File: java/leetcode/dp/delete_two_strings.java
/**
 * 583. Delete Operation for Two Strings
 * Given two words word1 and word2, find the minimum number of steps required to make word1 and word2 the same, where in each step you can delete one character in either string.

Example 1:

Input: "sea", "eat"
Output: 2
Explanation: You need one step to make "sea" to "ea" and another step to make "eat" to "ea".
Note:

The length of given words won't exceed 500.
Characters in given words can only be lower-case letters.
 */

public class delete_two_strings {

    // same as longest common substring between two strings
    public int minDistance(String word1, String word2) {
        int m = word1.length(), n = word2.length();
        int[][] dp = new int[m+1][n+1];
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (word1.charAt(i-1) == word2.charAt(j-1)) {
                    dp[i][j] = dp[i-1][j-1] + 1;
                } else {
                    dp[i][j] = Math.max(dp[i][j-1], dp[i-1][j]);
                }
            }
        }
        return m+n - 2 * dp[m][n];
    }
}

File: golang/leetcode/dp/delete_two_strings.go
package dp

// 583. Delete Operation for Two Strings

func minDistance(word1, word2 string) int {
	// similar as longest common subsequence
	// find lcs and then min deletion = m + n - 2 * lcs
	m, n := len(word1), len(word2)
	dp := make([][]int, m+1)
	for i := range dp {
		dp[i] = make([]int, n+1)
	}
	for i := 1; i <= m; i++ {
		for j := 1; j <= n; j++ {
			if word1[i-1] == word2[j-1] {
				dp[i][j] = dp[i-1][j-1] + 1
			} else {
				dp[i][j] = Max(dp[i][j-1], dp[i-1][j])
			}
		}
	}
	return m + n - 2*dp[m][n]
}

/*
func Max(x, y int) int {
	if x > y {
		return x
	}
	return y
}
*/
